### ISSUE TITLE
Yolo-NAS+Deepface

### ISSUE THREAD

**Issue by oussaifi-majdi** [serengil/deepface#822]
Hi Serengil,

Thanks for maintaining this amazing project. It has simplified many areas in facial verification/recognition.

i have a project face recognition and i want to use yolo-nas for detection and deepface for verification but can't find the correct code can you advice me 

thanks

that's my code :

import torch
from super_gradients.training import models
from super_gradients.training.models.detection_models.pp_yolo_e import PPYoloEPostPredictionCallback
from super_gradients.common.object_names import Models
from super_gradients.training import models
from super_gradients.common.object_names import Models
from super_gradients.training.utils.media.video import load_video
import numpy as np
import os
import cv2
from deepface import DeepFace 
from deepface.commons import distance as dst

rgb = r'C:\pfe_interdit\teste.jpg'
img_url = cv2.imread(rgb)
best_model = models.get(Models.YOLO_NAS_M, num_classes=1, checkpoint_path=r"C:\Users\Lenovo\PycharmProjects\yolo8\average_model.pth")
model = best_model.to('cuda' if torch.cuda.is_available() else "cpu")
# Load the known images and their encodings for face recognition
encodings = []
names = []
image_dir = r'C:\Users\Lenovo\PycharmProjects\images'
for root, dirs, files in os.walk(image_dir):
    for file in files:
        path = os.path.join(root, file)
        name = os.path.splitext(file)[0]
        encodings_person = DeepFace.represent(path, model_name="Facenet",detector_backend="skip") # since it can have several faces in the image we would like to take the first
        if len(encodings_person) > 0:
            encoding = encodings_person[0]
        # Add it to my Array
        encodings.append(encoding) # add the known face enconding
        names.append(name) # add the known face name 
# Predict using the model
results = model.predict(img_url)

for image_prediction in results:
    bboxes = image_prediction.prediction.bboxes_xyxy
for bbox in bboxes:
    x1, y1, x2, y2 = map(int, bbox)
    face_img = img_url[y1:y2, x1:x2]
    face_encoding = DeepFace.represent(face_img,detector_backend="skip")
    if len(face_encoding) > 0:
        # Get the first face encoding
        face_encoding = face_encoding[0]
        # Perform face recognition by comparing encodings
        face_names = []
        tolerance = 0.55
        for known_encoding, known_name in zip(encodings, names):            
            res=(face_encoding['embedding']==known_encoding['embedding'])
            if res:
                face_names.append(known_name)
        if len(face_names) > 0:
            name = ', '.join(face_names)
        else:
            name = 'Unknown'
        # Draw bounding box and name on the frame
        cv2.rectangle(img_url, (x1, y1), (x2, y2), (0, 255, 0), 2)
        cv2.putText(img_url, name, (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (0, 255, 0), 2)

# Display the output image with face bounding boxes and names
cv2.imshow('Face Recognition', img_url)
cv2.waitKey(0)
cv2.destroyAllWindows()

**Comment by oussaifi-majdi** [comment]
when I use face recognition library it gives the names of the correct faces and the other Unknown but when I use deepface whatever the method of calculating the distance it assigns names to all known and unknown faces

**Comment by serengil** [comment]
Do not set detector backend to skip. It will decrease the accuracy dramatically.

**Comment by oussaifi-majdi** [comment]
i want to use yolo-nas for detection thats why i want to use deepface in verification and recognition so i change it with enforce_detection=False but i have the same problem

import torch
from super_gradients.training import models
from super_gradients.training.models.detection_models.pp_yolo_e import PPYoloEPostPredictionCallback
from super_gradients.common.object_names import Models
from super_gradients.training import models
from super_gradients.common.object_names import Models
from super_gradients.training.utils.media.video import load_video
import numpy as np
import os
import cv2
from deepface import DeepFace 
from deepface.commons import distance as dst
from scipy.spatial import distance
import matplotlib.pyplot as plt
from PIL import Image, ImageDraw
from sklearn.metrics.pairwise import cosine_similarity
import math
import random
import faiss
import time

def show_img(imgs: list, img_names: list) -> None:
    imgs_count = len(imgs)
    for i in range(imgs_count):
        ax = plt.subplot(1, imgs_count, i+1)
        ax.imshow(imgs[i])
        ax.set_title(img_names[i])
        ax.set_xticks([])
        ax.set_yticks([])
    plt.tight_layout(h_pad=3)
    plt.show()


rgb = r'C:\pfe_interdit\mpwin.jpg'
img_url = cv2.imread(rgb)
best_model = models.get(Models.YOLO_NAS_M, num_classes=1, checkpoint_path=r"C:\Users\Lenovo\PycharmProjects\yolo8\average_model.pth")
model = best_model.to('cuda' if torch.cuda.is_available() else "cpu")

# Load the known images and their encodings for face recognition
image_dir = r"C:/Users/Lenovo/PycharmProjects/images/"
files = []
representations = []
representation = []
for r, d, f in os.walk(image_dir):
    for file in f:
        if ".jpg" in file:
            exact_file = f"{r}/{file}"
            objs = DeepFace.represent(img_path=exact_file, model_name="Facenet", enforce_detection=False)
            for obj in objs:
                embedding = obj["embedding"]
                representations.append([file, embedding])
print(f"deepface unit test items have {len(representations)} images")
num_dimensions=128
for i in range(100):
    key = f"synthetic_{i}.jpg"
    vector = [random.gauss(-0.5, 0.5) for _ in range(num_dimensions)]
    representations.append([key, vector])
print(f"after adding synthetic data, we have {len(representations)} items")
embeddings = []
for key, value in representations:
    embeddings.append(value)
tic = time.time()
d=64
nlist = 100
m = 8
quantizer = faiss.IndexFlatL2(d) 
index = faiss.IndexIVFPQ(quantizer, d, nlist, m, 8)
index = faiss.IndexFlatL2(num_dimensions)
toc = time.time()
print(f"faiss instance is initialized in {toc-tic} seconds")
tic = time.time()
index.train(np.array(embeddings, dtype = "f"))
index.add(np.array(embeddings, dtype = "f"))
toc = time.time()
print(f"{len(embeddings)} vectors stored in faiss intance in {toc-tic} seconds") 
#show_img(imgs,names)
#save
#faiss.write_index(index,"vector.index")
# Predict using the model
results = model.predict(img_url)

for image_prediction in results:
    bboxes = image_prediction.prediction.bboxes_xyxy
for bbox in bboxes:
    x1, y1, x2, y2 = map(int, bbox)
    face_img = img_url[y1:y2, x1:x2]
    target_representation = DeepFace.represent(img_path = face_img, model_name = "Facenet",enforce_detection=False)[0]["embedding"]
    target_representation = np.array(target_representation, dtype='f')
    target_representation = np.expand_dims(target_representation, axis=0)
    k = 1
    tic = time.time()
    distances, neighbours = index.search(target_representation, k)
    toc = time.time()
    print(f"search completed in {toc - tic} in seconds")
    face_names = []
    threshold = 10
    for idx, neighbour in enumerate(neighbours[0]):
        val=neighbours[0]
        x=representations[val[0]][1]
        neighbour_name = representations[neighbour][0]
        cosine_similarity = dst.findEuclideanDistance(target_representation, x)
        if cosine_similarity <threshold:
            print(cosine_similarity)
            face_names.append(representations[val[0]][0])
        toc = time.time()
        print(f"search completed in {toc - tic} in seconds")
                
    if len(face_names) > 0:
            name = ', '.join(face_names)
    else:
            name = 'Unknown'
    # Draw bounding box and name on the frame
    cv2.rectangle(img_url, (x1, y1), (x2, y2), (0, 255, 0), 2)
    cv2.putText(img_url, name, (x1, y1 - 10), cv2.FONT_HERSHEY_SIMPLEX, 0.9, (0, 255, 0), 2)
    
# Display the output image with face bounding boxes and names
cv2.imshow('Face Recognition', img_url)
cv2.waitKey(0)
cv2.destroyAllWindows()

**Comment by serengil** [comment]
Set enforce_detection to True. Otherwise, you will have low score.
